### ISSUE TITLE
Support bidirectional type inference for lambdas when expected type is a callable protocol class

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#1296]
This is related to https://github.com/microsoft/pylance-release/issues/754

**Comment by decorator-factory** [comment]
It would also be useful to deduce argument types of a function when you're decorating it.
```py
def register(handler: Callable[[int, str], None]): 
    ...

@register
def function(number, string):
    ...
```
Here,  `number` and `string` could be inferred to be of type `int` and `str`. Seems related to this.

**Comment by erictraut** [comment]
I don't plan to support argument type inference for non-lambda functions. There is a valid syntax for annotating arguments in this case.

**Comment by erictraut** [comment]
This is now addressed in Pyright 1.1.96, which I just published. It will also be in the next published release of Pylance.
